File: apriltag_pose_estimator_py/apriltag_pose_estimator_py/multi_tag_pose_estimator.py
import copy

import cv2
import numpy as np


class MultiTagPoseEstimator:
    """
    Multi-AprilTag pose estimator for calculating target point poses
    Uses multiple AprilTags to improve pose estimation accuracy
    """

    # ========================================================
    # Initialization
    # ========================================================
    def __init__(
        self,
        list_marker_ids,
        list_marker_offsets,
        list_point_offsets,
        tag_size,
        camera_matrix,
        dist_coeffs,
        base_marker_id=None,
    ):
        """
        Initialize the MultiTagObjectPoseEstimator point.

        Parameters:
            list_marker_ids (list): List of tag IDs to be used
            list_marker_offsets (list): X, Y distances between markers
            list_point_offsets (list): List of point position offsets in Nx3 array format [[x1, y1, z1], [x2, y2, z2], ...]
            tag_size (float): Actual size of the tag (unit: m)
            camera_matrix (np.ndarray): Camera intrinsic parameter matrix
            dist_coeffs (np.ndarray): Camera distortion coefficients
            base_marker_id (int, optional): ID of the base marker
        """
        self.marker_ids = copy.deepcopy(list_marker_ids)
        self.marker_offset_x, self.marker_offset_y = list_marker_offsets
        self.point_offsets = copy.deepcopy(list_point_offsets)
        self.point_count = len(list_point_offsets)
        self.tag_size = tag_size
        self.camera_matrix = copy.deepcopy(camera_matrix)
        self.dist_coeffs = copy.deepcopy(dist_coeffs)
        self.base_marker_id = (
            base_marker_id if base_marker_id is not None else self.marker_ids[0]
        )

    # ========================================================
    # Coordinate Transformation Utilities
    # ========================================================
    def rvec_tvec_to_matrix(self, r, t):
        """
        Convert rotation vector and translation vector to a 4x4 homogeneous transformation matrix.

        Parameters:
            r (np.ndarray): Rotation vector of shape (3, 1), in Rodrigues format
            t (np.ndarray): Translation vector of shape (3, 1)

        Returns:
            T (np.ndarray): 4x4 homogeneous transformation matrix (rotation + translation)
        """
        R = cv2.Rodrigues(r)[0]
        T = np.array(
            [
                [R[0, 0], R[0, 1], R[0, 2], t[0][0]],  # Row 1
                [R[1, 0], R[1, 1], R[1, 2], t[1][0]],  # Row 2
                [R[2, 0], R[2, 1], R[2, 2], t[2][0]],  # Row 3
                [0.0, 0.0, 0.0, 1.0],  # Row 4 (homogeneous bottom row)
            ],
            dtype=np.float32,
        )
        return T

    # ========================================================
    # AprilTag Detection and Pose Estimation
    # ========================================================
    def detect(self, tags, img):
        """
        Detect markers in the image and estimate point poses relative to the base marker.

        Parameters:
            tags (list): List of detected AprilTag objects
            img (np.ndarray): Current image (for drawing purposes)

        Returns:
            success (bool): True if all required markers are detected and pose estimated
            point_T (list of np.ndarray): List of 4x4 transformation matrices for each point
            img (np.ndarray): Image with visualization (axes and projected points)
        """
        # 1. Make a dictionary for fast lookup: tag_id → tag
        tag_dict = {tag.tag_id: tag for tag in tags}

        # 2. Check if all required marker IDs are found in the detected tags
        if not all(marker_id in tag_dict for marker_id in self.marker_ids):
            return False, [], img  # Not all required markers detected

        # 3. Prepare 3D (world) and 2D (image) correspondence arrays
        num_markers = len(self.marker_ids)
        pt3D = np.zeros([num_markers * 4, 3])  # 4 corners per tag
        pt2D = np.zeros([num_markers * 4, 2])
        a = self.tag_size / 2.0  # Half tag size
        ox, oy = self.marker_offset_x, self.marker_offset_y
        base_index = self.marker_ids.index(
            self.base_marker_id
        )  # base marker offset index

        # 4. Fill in pt3D and pt2D by following marker ID order
        for i, marker_id in enumerate(self.marker_ids):
            dx = ox * (i - base_index)
            dy = oy * (i - base_index)

            # 4.1 Define world coordinates for the four corners of each marker
            pt3D[i * 4 + 0] = [-a + dx, a + dy, 0.0]  # top-left
            pt3D[i * 4 + 1] = [a + dx, a + dy, 0.0]  # top-right
            pt3D[i * 4 + 2] = [a + dx, -a + dy, 0.0]  # bottom-right
            pt3D[i * 4 + 3] = [-a + dx, -a + dy, 0.0]  # bottom-left

            # 4.2 Use the tag's detected corners (2D image points)
            corners = tag_dict[marker_id].corners
            pt2D[i * 4 : (i + 1) * 4] = corners

        # 5. Estimate camera pose using solvePnP
        success, rvec, tvec = cv2.solvePnP(
            pt3D, pt2D, self.camera_matrix, self.dist_coeffs
        )
        if not success:
            return False, [], img  # pose estimation failed

        # 6. Convert to 4x4 transformation matrix
        T = self.rvec_tvec_to_matrix(rvec, tvec)
        marker_rotation = T[:3, :3]
        marker_position = T[:3, 3]

        # 7. Compute point poses and project them onto the image
        point_T = []
        for i, offset in enumerate(self.point_offsets):
            offset_vec = np.array(offset, dtype=np.float32)  # shape (3,)
            obj_pos = marker_position + marker_rotation @ offset_vec

            # 7.1 Project the 3D point position into image plane
            pos_cam = obj_pos.astype(np.float32)
            img_point = self.camera_matrix @ pos_cam
            img_point /= img_point[2]

            # 7.2 Draw white circle with black border
            img = cv2.circle(
                img, (int(img_point[0]), int(img_point[1])), 5, (255, 255, 255), -1
            )  # Fill white circle
            img = cv2.circle(
                img, (int(img_point[0]), int(img_point[1])), 5, (0, 0, 0), 2
            )  # Draw black border

            # 7.3 Store the transformation matrix for this point
            obj_T = T.copy()
            obj_T[:3, 3] = obj_pos
            point_T.append(obj_T)

        # 8. Visualize the base marker pose using axes
        img = cv2.drawFrameAxes(
            img,
            self.camera_matrix,
            self.dist_coeffs,
            marker_rotation.astype(np.float64),
            marker_position.reshape(3, 1).astype(np.float64),
            self.tag_size,
        )

        return True, point_T, img

File: apriltag_pose_estimator/src/multi_tag_pose_estimator.cpp
#include "apriltag_pose_estimator/multi_tag_pose_estimator.hpp"

#include <algorithm>
#include <array>
#include <iostream>
#include <stdexcept>

namespace apriltag_pose_estimator {

// ========================================================================
// Constructor
// ========================================================================
MultiTagPoseEstimator::MultiTagPoseEstimator(
    const std::vector<int>&   marker_ids,
    const std::vector<float>& marker_offsets_6n,
    float                     tag_size,
    const cv::Mat&            camera_matrix,
    const cv::Mat&            dist_coeffs,
    int                       base_marker_id)
    : marker_ids_(marker_ids),
      tag_size_(tag_size),
      camera_matrix_(camera_matrix.clone()),
      dist_coeffs_(dist_coeffs.clone()) {
    // Set base marker ID
    if (base_marker_id == -1) {
        base_marker_id_ = marker_ids[0];
    } else {
        base_marker_id_ = base_marker_id;
    }

    // Find base marker index
    base_index_ = -1;
    for (size_t i = 0; i < marker_ids_.size(); i++) {
        if (marker_ids_[i] == base_marker_id_) {
            base_index_ = static_cast<int>(i);
            break;
        }
    }
    if (base_index_ < 0) {
        throw std::invalid_argument(
            "MultiTagPoseEstimator: base_marker_id not found in marker_ids");
    }

    // Validate flat list length: 6 values per marker
    const size_t N = marker_ids_.size();
    if (marker_offsets_6n.size() != N * 6) {
        throw std::invalid_argument(
            "MultiTagPoseEstimator: marker_offsets must have size 6 * marker_ids.size()");
    }

    // Compile per-marker T_base <- marker_i (rotation order: Rz * Ry * Rx)
    marker_T_base_to_marker_.clear();
    marker_T_base_to_marker_.reserve(N);
    for (size_t i = 0; i < N; i++) {
        float x  = marker_offsets_6n[i * 6 + 0];
        float y  = marker_offsets_6n[i * 6 + 1];
        float z  = marker_offsets_6n[i * 6 + 2];
        float r  = marker_offsets_6n[i * 6 + 3];
        float p  = marker_offsets_6n[i * 6 + 4];
        float yw = marker_offsets_6n[i * 6 + 5];

        // base_marker entry is identity by definition.
        if (static_cast<int>(i) == base_index_) {
            x = y = z = r = p = yw = 0.0f;
        }

        Eigen::AngleAxisf Rx(r, Eigen::Vector3f::UnitX());
        Eigen::AngleAxisf Ry(p, Eigen::Vector3f::UnitY());
        Eigen::AngleAxisf Rz(yw, Eigen::Vector3f::UnitZ());
        Eigen::Matrix3f   R = Rz.matrix() * Ry.matrix() * Rx.matrix();

        Eigen::Matrix4f T   = Eigen::Matrix4f::Identity();
        T.block<3, 3>(0, 0) = R;
        T.block<3, 1>(0, 3) = Eigen::Vector3f(x, y, z);
        marker_T_base_to_marker_.push_back(T);
    }
}

// ========================================================================
// Pose Estimation Functions
// ========================================================================
bool MultiTagPoseEstimator::estimate(
    const std::vector<TagDetection>& tags,
    cv::Mat&                         img,
    cv::Mat&                         rvec,
    cv::Mat&                         tvec,
    Eigen::Matrix4f&                 base_transform_out) {
    // 1. Make a dictionary for fast lookup: tag_id → tag
    std::map<int, TagDetection> tag_dict;
    for (const auto& tag : tags) {
        tag_dict[tag.id] = tag;
    }

    // 2. Check if all required marker IDs are found in the detected tags
    for (int marker_id : marker_ids_) {
        if (tag_dict.find(marker_id) == tag_dict.end()) {
            registerMiss();
            return false;  // Not all required markers detected
        }
    }

    // 3. Prepare 3D (world) and 2D (image) correspondence arrays
    size_t                   num_markers = marker_ids_.size();
    std::vector<cv::Point3f> pt3D;
    std::vector<cv::Point2f> pt2D;
    pt3D.reserve(num_markers * 4);  // Pre-allocate memory (4 corners per marker)
    pt2D.reserve(num_markers * 4);

    float a = tag_size_ / 2.0f;  // Half tag size

    // Local marker corners (each marker's own frame, z = 0).
    const std::array<Eigen::Vector4f, 4> local_corners = {
        Eigen::Vector4f(-a, +a, 0.0f, 1.0f),  // top-left
        Eigen::Vector4f(+a, +a, 0.0f, 1.0f),  // top-right
        Eigen::Vector4f(+a, -a, 0.0f, 1.0f),  // bottom-right
        Eigen::Vector4f(-a, -a, 0.0f, 1.0f),  // bottom-left
    };

    // 4. Fill in pt3D and pt2D by following marker ID order.
    //    pt3D is expressed in the base_marker frame so that solvePnP yields
    //    the camera-to-base_marker transform directly.
    for (size_t i = 0; i < marker_ids_.size(); i++) {
        const int              marker_id = marker_ids_[i];
        const Eigen::Matrix4f& T         = marker_T_base_to_marker_[i];

        // 4.1 Transform each local corner into base_marker frame
        for (const auto& c : local_corners) {
            Eigen::Vector4f p_base = T * c;
            pt3D.emplace_back(p_base.x(), p_base.y(), p_base.z());
        }

        // 4.2 Use the tag's detected corners (2D image points)
        const auto& corners = tag_dict[marker_id].corners;
        pt2D.insert(pt2D.end(), corners.begin(), corners.end());
    }

    // 5. Estimate camera pose using solvePnP (Using SOLVEPNP_ITERATIVE)
    bool success     = cv::solvePnP(pt3D, pt2D, camera_matrix_, dist_coeffs_, rvec, tvec);
    bool reused_pose = false;  // True if last pose is reused

    if (!success) {
        registerMiss();
        return false;  // Pose estimation failed
    }

    // 6. Reprojection Error-based Outlier Rejection
    // 추정된 pose로 3D 포인트를 이미지에 재투영하여 평균 거리를 계산하고,
    // 임계값(max_reproj_error_) 초과 시 비정상 프레임으로 판단하여 이전 유효 pose를 사용
    {
        // 6.1 Reproject 3D points to image plane
        std::vector<cv::Point2f> projected;
        cv::projectPoints(pt3D, rvec, tvec, camera_matrix_, dist_coeffs_, projected);

        // 6.2 Calculate mean reprojection error
        double totalErr = 0.0;
        for (size_t i = 0; i < pt2D.size(); ++i) {
            double dx = projected[i].x - pt2D[i].x;
            double dy = projected[i].y - pt2D[i].y;
            totalErr += std::sqrt(dx * dx + dy * dy);
        }
        double meanReprojErr = totalErr / pt2D.size();

        // 6.3 Outlier Rejection (reuse last pose up to max_pose_reuse_ frames)
        if (meanReprojErr > max_reproj_error_) {
            if (has_last_valid_pose_ && reuse_count_ < max_pose_reuse_) {
                rvec        = last_rvec_.clone();
                tvec        = last_tvec_.clone();
                reused_pose = true;
                reuse_count_++;
            } else {
                resetStaleState();
                return false;
            }
        } else {
            last_rvec_           = rvec.clone();
            last_tvec_           = tvec.clone();
            has_last_valid_pose_ = true;
            reuse_count_         = 0;
        }
    }

    // 7. Convert to 4x4 transformation matrix & SLERP 필터 적용
    Eigen::Matrix4f T_cam_2_marker          = rvecTvecToMatrix(rvec, tvec);
    Eigen::Matrix4f T_cam_2_marker_filtered = pose_filter_.filter(T_cam_2_marker);

    // 7.1 Update rvec/tvec to filtered values
    {
        cv::Mat R_f(3, 3, CV_64F);
        for (int r = 0; r < 3; r++)
            for (int c = 0; c < 3; c++)
                R_f.at<double>(r, c) = static_cast<double>(T_cam_2_marker_filtered(r, c));
        cv::Rodrigues(R_f, rvec);
        tvec.at<double>(0) = T_cam_2_marker_filtered(0, 3);
        tvec.at<double>(1) = T_cam_2_marker_filtered(1, 3);
        tvec.at<double>(2) = T_cam_2_marker_filtered(2, 3);
    }

    // 8. Output: filtered base marker pose (camera frame)
    base_transform_out = T_cam_2_marker_filtered;

    // 9. Visualize the base marker pose using axes (skip reused poses)
    if (!reused_pose) {
        cv::drawFrameAxes(img, camera_matrix_, dist_coeffs_, rvec, tvec, tag_size_);
    }

    miss_count_ = 0;
    return true;
}

void MultiTagPoseEstimator::resetStaleState() {
    has_last_valid_pose_ = false;
    reuse_count_         = 0;
    pose_filter_.reset();
}

void MultiTagPoseEstimator::registerMiss() {
    miss_count_++;
    // Reset filter/cache after a long detection gap
    if (miss_count_ >= max_miss_reset_) {
        resetStaleState();
    }
}

// ========================================================================
// Utility Functions
// ========================================================================
Eigen::Matrix4f MultiTagPoseEstimator::rvecTvecToMatrix(
    const cv::Mat& rvec,
    const cv::Mat& tvec) {
    cv::Mat R;
    cv::Rodrigues(rvec, R);

    Eigen::Matrix4f T = Eigen::Matrix4f::Identity();
    for (int row = 0; row < 3; row++) {
        for (int col = 0; col < 3; col++) {
            T(row, col) = R.at<double>(row, col);
        }
        T(row, 3) = tvec.at<double>(row, 0);
    }

    return T;
}

}  // namespace apriltag_pose_estimator
